### PR TITLE
Prevent native file update

### PR DESCRIPTION
**Pull request by WcaleNieWolny**
If update is not major (example `1.1.0` -> `2.0.0` is major) prevent changing native code.
This is ensured in the following way:

1. Find all `.java`, `.kt`, `.scala`, `.swift` files in node modules
2. create a hash of said files
3. compare these hash with hash from remote database, if there is any differences then disallow upload 

/claim #87

### CODE REVIEW

**Comment by WcaleNieWolny** [comment]
image

This should be fine, I tested this regex and the whole read `node_modules` operation only took 200 ms

**Comment by WcaleNieWolny** [comment]
If performance ever becomes a problem we can implement workers or some other solution. However, I believe performance is good enough for now

**Comment by WcaleNieWolny** [comment]
Here is the schema PR: https://github.com/Cap-go/capgo/pull/299

## **DO NOT MERGE THIS PR WITHOUT MERGING THIS SCHEMA PR FIRST**

File: src/bundle/upload.ts [hunk 1]
```diff
@@ -291,3 +299,78 @@ export const uploadDeprecatedCommand = async (apikey: string, options: Options)
...
+  nativeFilesSpinner.start('Calculating checksum of native files')
+  const nativeCodeLocal = await findAllNativeCode()
+  nativeFilesSpinner.stop(`Native code length: ${nativeCodeLocal.length}`);
+
+  const hashes = nativeCodeLocal.map(val => val.hash)
+
+  const {data: dataChannel, error: errorChannel } = await supabase
+    .from('channels')
+    .select(`
+      version (
```
Comment (riderx) [inline src/bundle/upload.ts:315]: you should get the last version uploaded not the channel

**Comment by riderx** [comment]
The point of the proposed solution was to have the pakages list and check version by version to allow display the issue and help users prevent they do something wrong.
You solution is very smart, but the sha should be link to each version by the cli
After that it's the update endpoint who should decide not the CLI

**Comment by WcaleNieWolny** [comment]
I believe storing the versions of packages instead of native files sha256 could go very wrong very fast. Let's say that someone uses npm, he will have package.json and some dependency lock file. Now let's say someone else uses yarn. He also will have package.json and some different lock file. If we just store versions of all packages then we have to deal with a lot of edge cases.

**Comment by WcaleNieWolny** [comment]
About the update url. That I do not understand. This check operation is quite expensive and doing it on the server is a waste of money. We would be paying for cpu usage that is not necessary as this operation can be done in the cli

**Comment by WcaleNieWolny** [comment]
Now, we also have a privacy issue. Let's say we store something like this:
`[{"package_name": "@capacitor/android", "version": "^4.4.0"}]`

This raises the issue that we are storing information about a project that the client might now want us to have.
We probably would have to store hash of both of these fields

File: src/bundle/upload.ts (hunk 1 above)
Comment (WcaleNieWolny) [reply src/bundle/upload.ts:315]: Well... this raises an interesting question. How do we handle manual updates? Let's say user wants to downgrade from `2.0.0` -> `1.0.5`. With this code he could do that, even tho the native files might be different and the whole downgrade would require new app store version. But if he does complete this downgrade and we get the latest version (not the channel as you suggested) the we would be getting the `2.0.0` version even tho the app would be on `1.0.5`.

I believe getting this from the channel is not a terrible solution

**Comment by riderx** [comment]
My idea was to ignore any lock and just keep the version from package.json
The problem we have is when we distribute update, not when upload. Since many users will have different version installed.
Blocking the upload don’t help

**Comment by WcaleNieWolny** [comment]
> The problem we have is when we distribute update

Well, perhaps but we have a mechanism to prevent this exact issue
```json
{"major":true,"message":"Cannot upgrade major version","error":"disable_auto_update_to_major","version":"2.0.0","old":"1.3.14"}
```

This should prevent updating native files as updating native files requires new major release

**Comment by WcaleNieWolny** [comment]
@riderx what is the status on this? Will you merge this or do you want me to do exactly as you said in the review?

**Comment by WcaleNieWolny** [comment]
@riderx changed as requested, added an option to skip this native files check

**Comment by riderx** [comment]
@WcaleNieWolny the problem is semver require the user to do it properly where the feedback i get is it's hard for team to do semver for many reason.
So it would be a second check

**Pull request closed without merging**